Common/include/toolboxes/ndflattener.hpp

```diff
@@ -0,0 +1,793 @@
+ * the rank, zone index, and marker index.
+ *
+ * # Form the local NdFlattener
+ * by a "recursive function" like this:
+ *
+ *     auto f_local =
+ *     make_pair( nZone, [=](unsigned long iZone){
+ *       return make_pair( (Geometry of iZone)->GetnMarker() , [=](unsigned long iMarker){
+ *         return YOUR_PROPERTY(iZone, iMarker);
+ *       });
+ *     });
+ *     NdFlattener<2> nd_local(f_local);
+ *
+ * It might be safer to explicitly capture what you need in the lambda expressions.
+ *
+ * You can also construct an NdFlattener without any arguments, and use NdFlattener::initialize(f_local).
+ *
+ * f_local is
+ *  - a pair of an (exclusive) upper bound for the first index and a function that maps the first index to
+ *  - a pair of an (exclusive) upper bound for the second index and a function that maps the second index to
+ *  - (iterate this if there are more layers)
+ *  - the value corresponding to these indices.
+ * The template parameter "2" is the number of indices. The data type (default: su2double) and the index type
+ * (default: unsigned long) are further optional template parameters.
+ *
+ * # Form the global NdFlattener
+ * by "collective communication" like this:
+ *
+ *     NdFlattener<3> nd_global(Nd_MPI_Environment(), &nd_local);
+ *
+ * nd_global's first index is the rank, then the indices of nd_local follow.
+ *
+ * The struct Nd_MPI_Environment contains the relevant information about the parallel environment. The default
+ * values are suitable to communicate su2double data. In order to communicate data of another MPI datatype,
+ * use e.g. Nd_MPI_Environment(MPI_UNSIGNED_LONG).
+ *
+ * You can also construct an NdFlattener without any arguments, and use NdFlattener::initialize(mpi_env, &nd_local).
+ *
+ * # Refreshing
+ * If only the data but not the indices (i.e. the structure of sublists' lengths) have changed, you can refresh
+ * them with the call
+ *
+ *     nd_local.refresh(f_local); // , or
+ *     nd_global(Nd_MPI_Environment(), &nd_local); // respectively.
+ *
+ * An NdFlattener constructed or initialized from a "recursive function" or "collective communication" must be
+ * refreshed in the same way.
+ *
+ * We also provide a function NdFlattener::initialize_or_refresh, that initializes if not initialized, and refreshes
+ * otherwise.
+ *
+ * # Look-up
+ * You can access the NdFlattener via a "[]...[]" interface that resembles multidimensional arrays
+ * or pointer-to-pointer-... arrays. If you provide all of the K indices, this returns a reference
+ * to the data element. If the NdFlattener is declared const, the reference is also const. If you
+ * provide less than K indices, you obtain an IndexAccumulator object to which you can later pass more
+ * indices. With respect to the above example:
+ *
+ *     std::cout << nd_global[rank][zone][marker];
+ *     auto nd_g_rank = nd_global[rank];
+ *     nd_g_rank[zone][marker] += 1.0;
+ *
```

Comment:
I added a second output line with example value. This does also demonstrate that writing to the IndexAccumulator does not invalidate it (in case somebody asks).